rita/src/rita_common/dao_manager/mod.rs

```diff
@@ -0,0 +1,257 @@
+// Manages subnet DAO membership, DAOManager mantains a cache of subnet DAO entries.
+// The workflow goes as follows, an actor message DAOCheck is sent to DAOManager
+// if the identity is not on the DAO it will run a callback to tunnel manager to remove
+// that tunnel from operation. If the identity is on the DAO it will do nothing.
+// Entires from the DAO are cached for a configurable amount of time, this may of course
+// have the effect of adding someone to the DAO taking time to kick in.
+
+use actix::prelude::*;
+use actix_web::client::Connection;
+use actix_web::error::JsonPayloadError;
+use actix_web::*;
+use futures::Future;
+use std::collections::HashMap;
+use std::net::IpAddr;
+use std::net::SocketAddr;
+use std::time::Duration;
+use std::time::Instant;
+use tokio::net::TcpStream as TokioTcpStream;
+
+use althea_types::EthAddress;
+use althea_types::Identity;
+use num256::Uint256;
+use settings::RitaCommonSettings;
+
+use SETTING;
+
+// A json object specifcally for the web3 function
+// call response we expect from the SubnetDAO contract
+#[derive(Deserialize, Debug)]
+struct Web3Response {
+    jsonrpc: String,
+    id: u32,
+    result: Uint256,
+}
+
+pub struct DAOManager {
+    entries: HashMap<Identity, Vec<DAOEntry>>,
+}
+
+impl Actor for DAOManager {
+    type Context = Context<Self>;
+}
+impl Supervised for DAOManager {}
+impl SystemService for DAOManager {
+    fn service_started(&mut self, _ctx: &mut Context<Self>) {
+        info!("Tunnel manager started");
+    }
+}
+
+impl Default for DAOManager {
+    fn default() -> DAOManager {
+        DAOManager::new()
+    }
+}
+
+impl DAOManager {
+    fn new() -> DAOManager {
+        DAOManager {
+            entries: HashMap::<Identity, Vec<DAOEntry>>::new(),
+        }
+    }
+}
+
+pub struct DAOEntry {
+    on_list: bool,
+    dao_address: EthAddress,
+    id: Identity,
+    last_updated: Instant,
+}
+
+pub struct DAOCheck(Identity);
+impl Message for DAOCheck {
+    type Result = ();
+}
+
+impl Handler<DAOCheck> for DAOManager {
+    type Result = ();
+
+    fn handle(&mut self, msg: DAOCheck, _: &mut Context<Self>) -> Self::Result {
+        let their_id = msg.0;
+        check_cache(their_id, &self.entries);
+    }
+}
+
+/// Called by returning DAO requests, sends a message to TunnelManager
+pub struct CacheCallback((Identity, EthAddress, Web3Response));
```

Comment:
Its probably my favorite nitpick but in non-trivial messages I'd rather see named structs instead of tuples. This way we can avoid mysterious calls to `(msg.0).2` etc.

Reply:
I'm of the same opinion, I guess I'll make a cache callback object